python/cudf/cudf/core/index.py

```diff
@@ -2744,6 +2748,212 @@ def categories(self):
+        start = start + freq
+        right_col = cupy.arange(start, end, freq)
+    elif start is not None and end is not None:
+        left_col = cupy.arange(start, end, freq)
+        start = start + 1
+        end = end + 1
+        right_col = cupy.arange(start, end, freq)
+    else:
+        raise ValueError(
+            "Of the four parameters: start, end, periods, and "
+            "freq, atleast two must be specified"
+        )
+    if len(right_col) == 0 or len(left_col) == 0:
+        return cudf.IntervalIndex([], closed=closed)
+
+    interval_col = column.build_interval_column(
+        left_col, right_col, closed=closed
+    )
+    return IntervalIndex(interval_col)
+
+
+class IntervalIndex(GenericIndex):
+    """
+    Immutable index of intervals that are closed on the same side.
+
+    Parameters
+    data : array-like (1-dimensional)
+    Array-like containing Interval objects from which to build the
+    IntervalIndex.
+
+    closed : {"left", "right", "both", "neither"}, default "right"
+    Whether the intervals are closed on the left-side, right-side,
+    both or neither.
+
+    dtype : dtype or None, default None
+    If None, dtype will be inferred.
+
+    copy : bool, default False
+    Copy the input data.
+
+    name : object, optional
+    Name to be stored in the index.
+
+    Returns
+    IntervalIndex
+    """
+
+    def __new__(
+        cls, data=None, closed=None, dtype=None, copy=False, name=None,
+    ) -> "IntervalIndex":
+        if copy:
+            data = column.as_column(data, dtype=dtype).copy(deep=True)
+        out = Frame.__new__(cls)
+        kwargs = _setdefault_name(data, name=name)
+        if isinstance(data, IntervalColumn):
+            data = data
+        elif isinstance(data, pd.Series) and (is_interval_dtype(data.dtype)):
+            data = column.as_column(pa.Array.from_pandas(data), dtype=dtype,)
+        elif isinstance(data, (pd._libs.interval.Interval, pd.IntervalIndex)):
+            data = column.as_column(data, dtype=dtype,)
+        else:
+            if data is not None and data != [] and data[0].closed != closed:
+                # when closed is not the same as the data's closed
+                # we need to change the data closed value,
+                left_col = [data[i].left for i in range(len(data))]
+                # creating a col out of the left child so we can get
+                # the correct dtype
+                left = column.as_column(left_col)
+                data = column.as_column(
+                    data,
+                    dtype=IntervalDtype(left.dtype, closed=closed)
+                    if dtype is None
+                    else dtype,
+                )
+            if not data and closed != "right":
+                data = column.build_interval_column([], [], closed=closed)
+            else:
+                data = column.as_column(
+                    data, dtype="interval" if dtype is None else dtype
+                )
```

Comment:
Can you describe the situation this `else` block is trying to handle here? Looks like a list of pandas interval objects?

Reply:
The else block catches the case where a user puts in an empty list but sets closed to a value. It makes sure that the closed value is correct even if the index is empty. For example in the below case 

```
 >>>example_interval= IntervalIndex([],
              closed='right',
              dtype='interval[int64]'), right = Int64Index([], dtype='int64')

... IntervalIndex([],
               closed='right',
               dtype='interval[int64]')
```
I should probably be more explicit here TBF

Reply:
Realising you were probably referring to the whole `else` block and not just the one at the end 😅 
I've changed the code so that it's not all under one else block and is more explicit.
 Its mainly trying to handle three cases:
1. Where IntervalIndex receives two different `closed` values e.g. `IntervalIndex([Interval(30, 50, closed='right')], closed = 'left')`. We need to be able to make the final `closed` value `left`. The first part of the statement tries to do this. It is a bit messy, so very open to feedback on it and will think through it a bit more myself. 
2. When the user inputs an empty list into the IntervalIndex and also adds a `closed` value to it. 
3. Other cases, like when the input is a list of intervals and possibly other edge cases.